enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/developer/BearerTokenCommand.java

```diff
@@ -0,0 +1,140 @@
+import com.oracle.bmc.http.client.HttpResponse;
+import com.oracle.bmc.http.client.Method;
+import com.oracle.bmc.http.client.jersey.JerseyHttpProvider;
+import com.oracle.bmc.http.signing.RequestSigningFilter;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import javax.ws.rs.core.MediaType;
+import org.netbeans.modules.cloud.oracle.OCIManager;
+import org.netbeans.modules.cloud.oracle.OCIProfile;
+import org.netbeans.modules.cloud.oracle.assets.TempFileGenerator;
+import org.netbeans.spi.lsp.CommandProvider;
+import org.openide.util.RequestProcessor;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ *
+ * @author Jan Horvath
+ */
+@ServiceProvider(service = CommandProvider.class)
+public class BearerTokenCommand implements CommandProvider {
+
+    private static final String COMMAND_CREATE_BEARER_TOKEN = "nbls.cloud.assets.createBearerToken"; //NOI18N
+
+    private static final String API_VERSION = "20180419"; //NOI18N
+    private static String lastPath;
+    private static Instant lastTaken = Instant.MIN;
+
+    private static final RequestProcessor RP = new RequestProcessor(BearerTokenCommand.class);
+
+    private static final Set COMMANDS = new HashSet<>(Arrays.asList(
+            COMMAND_CREATE_BEARER_TOKEN
+    ));
+
+    @Override
+    public Set<String> getCommands() {
+        return Collections.unmodifiableSet(COMMANDS);
+    }
+
+    @Override
+    public CompletableFuture<Object> runCommand(String command, List<Object> arguments) {
+        final String ocirServer;
+        if (arguments != null && !arguments.isEmpty()) {
+            JsonPrimitive s = (JsonPrimitive) arguments.get(0);
+            ocirServer = s.getAsString();
+        } else {
+            return null;
+        }
+        CompletableFuture<Object> result = new CompletableFuture<>();
+        RP.post(() -> {
+            OCIProfile session = OCIManager.getDefault().getActiveProfile();
+            try {
+                result.complete(generateBearerToken(session, ocirServer));
+            } catch (URISyntaxException | InterruptedException | ExecutionException | IOException e) {
+                result.completeExceptionally(e);
+            }
+        });
+        return result;
+    }
+
+    public static synchronized String generateBearerToken(OCIProfile provider, String ocirServer) throws URISyntaxException, IOException, ExecutionException, InterruptedException  {
+        if (Duration.between(lastTaken, Instant.now()).toMinutes() > 55) {
```

Comment:
Q: since the cache file's timestamp is +- the same as `lastTaken`, could it be taken after NBLS start as a valid token - or is it preferred to generate a new token after each NBLS restart ?

Reply:
I prefer to keep the logic as it is now. It is cheap to obtain the token.